In [37]:
from utils import setup_env_path

setup_env_path()

%load_ext autoreload
%autoreload 2

Former working directory:  /Users/ambroisebertin/Desktop/prog/prog_abeilles/fil-rouge-pollinisateurs
Current working directory:  /Users/ambroisebertin/Desktop/prog/prog_abeilles/fil-rouge-pollinisateurs
The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [38]:
import pandas as pd
import numpy as np

In [39]:
API_keys = [
    "2b10Me1HF0rfjoGWCseolNa8e",
    "2b10Xs7brWPuBdRTSeWx9V7HJu",
    "2b10umm8L2jTYjlWrJKPfshJ0O",
    "2b10sEg5pSsrtVT372XqlrzcLe",
    "2b10LDWWlXJwuS6FleaqfK6Rke",
]

## Préparation des données

In [40]:
# Import spipoll.csv as a pandas dataframe
spipoll = pd.read_csv("././data/spipoll.csv",low_memory=False)

# Extract relevant columns from the spipoll dataframe
plantes = spipoll[['collection_id', 'plante_sc', 'plante_fr',
       'plante_precision', 'plante_inconnue', 'plante_caractere',
       'photo_fleur', 'photo_plante', 'photo_feuille']]

# Shrinking the data by grouping by collection_id
# Keep only the first row for each unique value of collection_id
plantes = plantes.drop_duplicates(subset='collection_id', keep='first')

plantes.shape

(75372, 9)

In [41]:
values_to_check = ["Je ne sais pas", "Plante inconnue"]

In [42]:
plantes_sc_unlabelled = plantes.loc[ #renommer en plantes_unlabelled_sc pour plus de clarté ??
    (
        (plantes["plante_sc"].isna() | plantes["plante_sc"].isin(values_to_check))
    )
    | ((plantes["plante_inconnue"] == 1.0))
]

# creation du dataframe contenant les valeurs de plantes mais pas de plantes_sc_unlabelled
plantes_sc_labelled = plantes[~plantes.index.isin(plantes_sc_unlabelled.index)]

# in plante add "1" to the column group for each row that is in plantes_sc_labelled
plantes.loc[plantes_sc_labelled.index, 'group'] = 1

plantes_sc_labelled.shape

(63465, 9)

In [43]:
# verification des dimensions des deux df par rapport à la dim de plantes
plantes_sc_labelled.shape[0] + plantes_sc_unlabelled.shape[0] == plantes.shape[0]

True

Compression de l'information étiquettée redondante :

In [44]:
# group plantes_sc_labelled by unique values of plante_sc
plantes_sc_labelled = plantes_sc_labelled.drop_duplicates(subset='plante_sc', keep='first')

# display the dimension of the dataframe
plantes_sc_labelled.shape

(2620, 9)

On a donc un df avec 2620 plantes dont le nom est spécifié dans plante_sc et 11.907 (~1.8% des données d'origine, spipoll.csv) dont le nom n'est pas spécifié dans plante_sc.

In [45]:
# ajoute trois colonnes après plante_sc dans le dataframe plantes_sc_labelled "Famille", "Genre" et "Espèce" entre la colonne plante_sc et plante_fr
plantes_sc_labelled.insert(2, "Famille", np.nan)
plantes_sc_labelled.insert(3, "Genre", np.nan)
plantes_sc_labelled.insert(4, "Espece", np.nan)

# ajoute trois colonnes après plante_sc dans le dataframe plantes_sc_unlabelled "Famille", "Genre" et "Espèce" entre la colonne plante_sc et plante_fr
plantes_sc_unlabelled.insert(2, "Famille", np.nan)
plantes_sc_unlabelled.insert(3, "Genre", np.nan)
plantes_sc_unlabelled.insert(4, "Espece", np.nan)

In [46]:
plantes_sc_labelled.head()

collection_id                                          plante_sc  Famille  \
0              1  Les Trèfles à fleurs blanches ou roses en boul...      NaN   
2              2                                    Acanthus mollis      NaN   
5              3  Les Orchidées à fleurs blanches, jaunes ou ver...      NaN   
6              4  Les Vipérines à fleurs bleues-violacées (des e...      NaN   
10             5                        Les Pâquerettes (Bellis sp)      NaN   

    Genre  Espece      plante_fr                       plante_precision  \
0     NaN     NaN            NaN                                    NaN   
2     NaN     NaN  Acanthe molle                                    NaN   
5     NaN     NaN            NaN  orchis bouc - Himantoglossum hircinum   
6     NaN     NaN            NaN                               Theaceae   
10    NaN     NaN            NaN                                    NaN   

    plante_inconnue plante_caractere  \
0               0.0        spontanée   
2               NaN          plantée   
5               0.0        spontanée   
6               0.0          plantée   
10              0.0        spontanée   

                                          photo_fleur  \
0   https://spgp-api.65mo.fr/api/containers/spgp/d...   
2   https://spgp-api.65mo.fr/api/containers/spgp/d...   
5   https://spgp-api.65mo.fr/api/containers/spgp/d...   
6   https://spgp-api.65mo.fr/api/containers/spgp/d...   
10  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                         photo_plante  \
0   https://spgp-api.65mo.fr/api/containers/spgp/d...   
2   https://spgp-api.65mo.fr/api/containers/spgp/d...   
5   https://spgp-api.65mo.fr/api/containers/spgp/d...   
6   https://spgp-api.65mo.fr/api/containers/spgp/d...   
10  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                        photo_feuille  
0   https://spgp-api.65mo.fr/api/containers/spgp/d...  
2   https://spgp-api.65mo.fr/api/containers/spgp/d...  
5   https://spgp-api.65mo.fr/api/containers/spgp/d...  
6   https://spgp-api.65mo.fr/api/containers/spgp/d...  
10  https://spgp-api.65mo.fr/api/containers/spgp/d...

## Appel à l'API

In [47]:
from data_quality.plant_treatment.plantnet_api import PlantNetPredictor

def call_API(iloc, api_key_index=0):
    predictor = PlantNetPredictor(key = API_keys[api_key_index])
    prediction = None

    try:
        prediction = predictor.predict(imageURL= plantes_sc_labelled['photo_fleur'].iloc[iloc],
                        organs = "auto",
                        includeRelatedImages=False)
    except Exception as e:
        print(f"Error: {e}")
        if api_key_index + 1 < len(API_keys):
            return get_plant_details(iloc, api_key_index + 1)
        else:
            print("You used all your API keys for today.")
            return None

    famille = prediction["results"][0]["species"]["family"]["scientificNameWithoutAuthor"] # famille
    genre = prediction["results"][0]["species"]["genus"]["scientificNameWithoutAuthor"] # genre
    espece = prediction["results"][0]["species"]["scientificNameWithoutAuthor"] # espece

    return [famille, genre, espece]

In [48]:
# appelle la fonction call API sur les 10 premières lignes de plantes_sc_labelled et affiche toutes les sorties
for i in range(10):
    print(call_API(i))

['Fabaceae', 'Trifolium', 'Trifolium pratense']
['Acanthaceae', 'Acanthus', 'Acanthus spinosus']
['Orchidaceae', 'Himantoglossum', 'Himantoglossum hircinum']
['Theaceae', 'Camellia', 'Camellia sasanqua']
['Asteraceae', 'Bellis', 'Bellis sylvestris']
['Ericaceae', 'Erica', 'Erica x darleyensis']
['Fabaceae', 'Adenocarpus', 'Adenocarpus foliolosus']
['Asteraceae', 'Taraxacum', 'Taraxacum palustre']
['Brassicaceae', 'Sinapis', 'Sinapis arvensis']
['Asteraceae', 'Senecio', 'Senecio inaequidens']


In [50]:
plantes_sc_labelled.head()

collection_id                                          plante_sc   Famille  \
0              1  Les Trèfles à fleurs blanches ou roses en boul...  Fabaceae   
2              2                                    Acanthus mollis       NaN   
5              3  Les Orchidées à fleurs blanches, jaunes ou ver...       NaN   
6              4  Les Vipérines à fleurs bleues-violacées (des e...       NaN   
10             5                        Les Pâquerettes (Bellis sp)       NaN   

        Genre              Espece      plante_fr  \
0   Trifolium  Trifolium pratense            NaN   
2         NaN                 NaN  Acanthe molle   
5         NaN                 NaN            NaN   
6         NaN                 NaN            NaN   
10        NaN                 NaN            NaN   

                         plante_precision  plante_inconnue plante_caractere  \
0                                     NaN              0.0        spontanée   
2                                     NaN              NaN          plantée   
5   orchis bouc - Himantoglossum hircinum              0.0        spontanée   
6                                Theaceae              0.0          plantée   
10                                    NaN              0.0        spontanée   

                                          photo_fleur  \
0   https://spgp-api.65mo.fr/api/containers/spgp/d...   
2   https://spgp-api.65mo.fr/api/containers/spgp/d...   
5   https://spgp-api.65mo.fr/api/containers/spgp/d...   
6   https://spgp-api.65mo.fr/api/containers/spgp/d...   
10  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                         photo_plante  \
0   https://spgp-api.65mo.fr/api/containers/spgp/d...   
2   https://spgp-api.65mo.fr/api/containers/spgp/d...   
5   https://spgp-api.65mo.fr/api/containers/spgp/d...   
6   https://spgp-api.65mo.fr/api/containers/spgp/d...   
10  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                        photo_feuille  
0   https://spgp-api.65mo.fr/api/containers/spgp/d...  
2   https://spgp-api.65mo.fr/api/containers/spgp/d...  
5   https://spgp-api.65mo.fr/api/containers/spgp/d...  
6   https://spgp-api.65mo.fr/api/containers/spgp/d...  
10  https://spgp-api.65mo.fr/api/containers/spgp/d...

In [58]:
from tqdm import tqdm

def remplir_tableau(df, num_rows):
    df = df.reset_index(drop=True)
    # parcours toutes les lignes de df et appelle la fonction call_API sur chaque ligne vide et stocke les résultats dans les colonnes "Famille", "Genre", et "Espèce"
    for i in tqdm(range(num_rows)):
        if pd.isnull(df.loc[i, "Famille"]) or pd.isnull(df.loc[i, "Genre"]) or pd.isnull(df.loc[i, "Espece"]):
            try:
                df.loc[i, ["Famille", "Genre", "Espece"]] = call_API(i)
            except Exception as e:
                print(f"Error: {e}")
                continue
    return df

remplir_tableau(plantes_sc_labelled, 8)

100%|██████████| 8/8 [00:20<00:00,  2.59s/it]


collection_id                                          plante_sc  \
0                1  Les Trèfles à fleurs blanches ou roses en boul...   
1                2                                    Acanthus mollis   
2                3  Les Orchidées à fleurs blanches, jaunes ou ver...   
3                4  Les Vipérines à fleurs bleues-violacées (des e...   
4                5                        Les Pâquerettes (Bellis sp)   
...            ...                                                ...   
2615        75 468                                  Arbutus andrachne   
2616        75 525                                    Pavonia hastata   
2617        75 554                                   Medicago falcata   
2618        75 625                             Helianthus divaricatus   
2619        75 657                                  Sideritis hirsuta   

          Famille           Genre                   Espece  \
0        Fabaceae       Trifolium       Trifolium pratense   
1     Acanthaceae        Acanthus        Acanthus spinosus   
2     Orchidaceae  Himantoglossum  Himantoglossum hircinum   
3        Theaceae        Camellia        Camellia sasanqua   
4      Asteraceae          Bellis        Bellis sylvestris   
...           ...             ...                      ...   
2615          NaN             NaN                      NaN   
2616          NaN             NaN                      NaN   
2617          NaN             NaN                      NaN   
2618          NaN             NaN                      NaN   
2619          NaN             NaN                      NaN   

                         plante_fr                       plante_precision  \
0                              NaN                                    NaN   
1                    Acanthe molle                                    NaN   
2                              NaN  orchis bouc - Himantoglossum hircinum   
3                              NaN                               Theaceae   
4                              NaN                                    NaN   
...                            ...                                    ...   
2615           Arbousier de Chypre                                    NaN   
2616     Pavonie (Pavonia hastata)                                    NaN   
2617           Luzerne en faucille                                    NaN   
2618  Hélianthe à feuilles Ètalées                                    NaN   
2619           Crapaudine d'Endres                                    NaN   

      plante_inconnue plante_caractere  \
0                 0.0        spontanée   
1                 NaN          plantée   
2                 0.0        spontanée   
3                 0.0          plantée   
4                 0.0        spontanée   
...               ...              ...   
2615              0.0        spontanée   
2616              NaN          plantée   
2617              0.0        spontanée   
2618              NaN        spontanée   
2619              NaN          plantée   

                                            photo_fleur  \
0     https://spgp-api.65mo.fr/api/containers/spgp/d...   
1     https://spgp-api.65mo.fr/api/containers/spgp/d...   
2     https://spgp-api.65mo.fr/api/containers/spgp/d...   
3     https://spgp-api.65mo.fr/api/containers/spgp/d...   
4     https://spgp-api.65mo.fr/api/containers/spgp/d...   
...                                                 ...   
2615  https://spgp-api.65mo.fr/api/containers/spgp/d...   
2616  https://spgp-api.65mo.fr/api/containers/spgp/d...   
2617  https://spgp-api.65mo.fr/api/containers/spgp/d...   
2618  https://spgp-api.65mo.fr/api/containers/spgp/d...   
2619  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                           photo_plante  \
0     https://spgp-api.65mo.fr/api/containers/spgp/d...   
1     https://spgp-api.65mo.fr/api/containers/spgp/d...   
2     https://spgp-api.65mo.fr/api/containers/spgp/d...   
3     ht

In [61]:
# affiche les 10 premières lignes de plantes_sc_labelled
plantes_sc_labelled.head(10)


collection_id                                          plante_sc   Famille  \
0              1  Les Trèfles à fleurs blanches ou roses en boul...  Fabaceae   
2              2                                    Acanthus mollis       NaN   
5              3  Les Orchidées à fleurs blanches, jaunes ou ver...       NaN   
6              4  Les Vipérines à fleurs bleues-violacées (des e...       NaN   
10             5                        Les Pâquerettes (Bellis sp)       NaN   
24             6  Les Bruyères à fleurs roses en grappes (des es...       NaN   
25             7             Les Ajoncs (des espèces du genre Ulex)       NaN   
41            12                      Les Pissenlits (Taraxacum sp)       NaN   
47            15  Les Crucifères jaunes (des espèces de la famil...       NaN   
54            17                                   Senecio jacobaea       NaN   

        Genre              Espece           plante_fr  \
0   Trifolium  Trifolium pratense                 NaN   
2         NaN                 NaN       Acanthe molle   
5         NaN                 NaN                 NaN   
6         NaN                 NaN                 NaN   
10        NaN                 NaN                 NaN   
24        NaN                 NaN                 NaN   
25        NaN                 NaN                 NaN   
41        NaN                 NaN                 NaN   
47        NaN                 NaN                 NaN   
54        NaN                 NaN  Le Sénéçon jacobée   

                         plante_precision  plante_inconnue  \
0                                     NaN              0.0   
2                                     NaN              NaN   
5   orchis bouc - Himantoglossum hircinum              0.0   
6                                Theaceae              0.0   
10                                    NaN              0.0   
24          Bruyère carnée (Erica carnea)              0.0   
25                 Ajonc (Ulex europaeus)              0.0   
41       Pissenlit (Taraxacum officinale)              0.0   
47                                    NaN              0.0   
54                                    NaN              0.0   

      plante_caractere                                        photo_fleur  \
0            spontanée  https://spgp-api.65mo.fr/api/containers/spgp/d...   
2              plantée  https://spgp-api.65mo.fr/api/containers/spgp/d...   
5            spontanée  https://spgp-api.65mo.fr/api/containers/spgp/d...   
6              plantée  https://spgp-api.65mo.fr/api/containers/spgp/d...   
10           spontanée  https://spgp-api.65mo.fr/api/containers/spgp/d...   
24             plantée  https://spgp-api.65mo.fr/api/containers/spgp/d...   
25  ne se prononce pas  https://spgp-api.65mo.fr/api/containers/spgp/d...   
41           spontanée  https://spgp-api.65mo.fr/api/containers/spgp/d...   
47           spontanée  https://spgp-api.65mo.fr/api/containers/spgp/d...   
54           spontanée  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                         photo_plante  \
0   https://spgp-api.65mo.fr/api/containers/spgp/d...   
2   https://spgp-api.65mo.fr/api/containers/spgp/d...   
5   https://spgp-api.65mo.fr/api/containers/spgp/d...   
6   https://spgp-api.65mo.fr/api/containers/spgp/d...   
10  https://spgp-api.65mo.fr/api/containers/spgp/d...   
24  https://spgp-api.65mo.fr/api/containers/spgp/d...   
25  https://spgp-api.65mo.fr/api/containers/spgp/d...   
41  https://spgp-api.65mo.fr/api/containers/spgp/d...   
47  https://spgp-api.65mo.fr/api/containers/spgp/d...   
54  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                        photo_feuille  
0   https://spgp-api.65mo.fr/api/containers/spgp/d...  
2   https://spgp-api.65mo.fr/api/containers/spgp/d...  
5   https://spgp-api.65mo.fr/api/containers/spgp/d...  
6   https://spgp-api.65mo.fr/api/containers/spgp/d...  
10  https://spgp-api.65mo.fr/api/containers/

In [60]:
plantes_sc_labelled.shape

(2620, 12)